File: app/src/main/java/piuk/blockchain/android/injection/ServiceModule.java
package piuk.blockchain.android.injection;

import com.blockchain.koin.KoinDaggerModule;
import dagger.Module;
import dagger.Provides;
import info.blockchain.wallet.ApiCode;
import info.blockchain.wallet.api.FeeApi;
import info.blockchain.wallet.api.WalletApi;
import info.blockchain.wallet.api.WalletExplorerEndpoints;
import info.blockchain.wallet.contacts.Contacts;
import info.blockchain.wallet.ethereum.EthAccountApi;
import info.blockchain.wallet.payment.Payment;
import info.blockchain.wallet.prices.PriceApi;
import info.blockchain.wallet.prices.PriceEndpoints;
import info.blockchain.wallet.settings.SettingsManager;
import info.blockchain.wallet.shapeshift.ShapeShiftApi;
import info.blockchain.wallet.shapeshift.ShapeShiftEndpoints;
import piuk.blockchain.android.data.fingerprint.FingerprintAuth;

@Module
class ServiceModule extends KoinDaggerModule {

    @Provides
    SettingsManager provideSettingsManager() {
        return get(SettingsManager.class);
    }

    @Provides
    Contacts provideContacts() {
        return get(Contacts.class);
    }

    @Provides
    WalletExplorerEndpoints provideWalletExplorerEndpoints() {
        return get(WalletExplorerEndpoints.class);
    }

    @Provides
    ShapeShiftEndpoints provideShapeShiftEndpoints() {
        return get(ShapeShiftEndpoints.class);
    }

    @Provides
    PriceEndpoints providePriceEndpoints() {
        return get(PriceEndpoints.class);
    }

    @Provides
    WalletApi provideWalletApi() {
        return get(WalletApi.class);
    }

    @Provides
    Payment providePayment() {
        return get(Payment.class);
    }

    @Provides
    FeeApi provideFeeApi() {
        return get(FeeApi.class);
    }

    @Provides
    ApiCode provideApiCode() {
        return get(ApiCode.class);
    }

    @Provides
    PriceApi providePriceApi() {
        return get(PriceApi.class);
    }

    @Provides
    ShapeShiftApi provideShapeShiftApi() {
        return get(ShapeShiftApi.class);
    }

    @Provides
    FingerprintAuth provideFingerprintAuth() {
        return get(FingerprintAuth.class);
    }

    @Provides
    EthAccountApi provideEthAccountApi() {
        return get(EthAccountApi.class);
    }
}

File: app/src/main/java/com/blockchain/koin/modules/serviceModule.kt
package com.blockchain.koin.modules

import info.blockchain.wallet.ApiCode
import info.blockchain.wallet.BlockchainFramework
import info.blockchain.wallet.api.FeeApi
import info.blockchain.wallet.api.FeeEndpoints
import info.blockchain.wallet.api.WalletApi
import info.blockchain.wallet.api.WalletExplorerEndpoints
import info.blockchain.wallet.ethereum.EthAccountApi
import info.blockchain.wallet.payment.Payment
import info.blockchain.wallet.settings.SettingsManager
import org.koin.dsl.module.applicationContext
import piuk.blockchain.android.data.fingerprint.FingerprintAuth
import piuk.blockchain.android.data.fingerprint.FingerprintAuthImpl
import retrofit2.Retrofit

val serviceModule = applicationContext {

    bean { SettingsManager(get()) }

    bean { get<Retrofit>("explorer").create(WalletExplorerEndpoints::class.java) }

    bean { get<Retrofit>("api").create(FeeEndpoints::class.java) }

    factory { WalletApi(get(), get()) }

    factory { Payment() }

    factory { FeeApi(get()) }

    factory {
        object : ApiCode {
            override val apiCode: String
                get() = BlockchainFramework.getApiCode()
        } as ApiCode
    }

    factory { FingerprintAuthImpl() as FingerprintAuth }

    factory { EthAccountApi() }
}
